# Conceptual



1. What is the difference between train and test set?

    A train set of data is used to train a model that will use the test data set to predict an outcome.
    
    
2. Why is it important to have a seperate data set for training?

    So that we can determine the predictive relatiionship between features when applied to a model, in order to assess the strength of the relationship when test data set is used against the same model.
    
    
3. What are some advantages and disadvantages of the KNN algorithm?
 
   Advantages:
    Training is fast
    Learn complex target functions
    Don't lose information

   Disadvantages:
    Slow at query time
    Lot's of storage
    Easily fooled by irrelevant features and high dimensions


4. Explain in your own words the bias variance tradeoff.
    
    Bias Variance trade off - 
    
5. Do you think KNN suffers from high bias or high variance?
    
    High Variance
    
    
6. List some possible solutions when suffering from high variance.
    
    Remove features or add more data.OR use another algorithm
    
    
7. List some possible solutions when suffereing from high bias.
    
    Add more features or use a different algorithm
    

8. Would you choose K when doing K Nearest Neighbors?
    
    No you use K Fold cross validation to determine the optimal k value

# Practical

##Questions
1. Perform regression on the prostate data set. Your goal is to minimize the mean squared error (MSE). Split this into a several parts and at a minimum utilize
    * Linear Regression
    * KNN
    * Lasso
    * Ridge 
    * Use cross validation to tune the best model you can

In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import linear_model
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler
from sklearn import cross_validation
from sklearn.metrics import mean_squared_error
from sklearn import grid_search
%matplotlib inline
from sklearn import neighbors
from sklearn.preprocessing import PolynomialFeatures
from sklearn import preprocessing


ds_cancer=pd.read_csv('C:/Users/sawadhawan/Documents/GitHub/DAT_ATL_15/Datasets/ProstateCancer.csv')
del ds_cancer['Unnamed: 0']

In [36]:
ds_cancer.head()

lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45      lpsa  \
0 -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0 -0.430783   
1 -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0 -0.162519   
2 -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20 -0.162519   
3 -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0 -0.162519   
4  0.751416  3.432373   62 -1.386294    0 -1.386294        6      0  0.371564   

  train  
0     T  
1     T  
2     T  
3     T  
4     T

In [37]:
ds_cancer_trn=ds_cancer[ds_cancer['train']=='T']
del ds_cancer_trn['train']
ds_cancer_test=ds_cancer[ds_cancer['train']=='F']
del ds_cancer_test['train']
x_train=ds_cancer_trn[['lcavol','lweight','age','lbph','svi','lcp','gleason','pgg45']]
y_train=ds_cancer_trn[['lpsa']]
x_test = ds_cancer_test[['lcavol','lweight','age','lbph','svi','lcp','gleason','pgg45']]
y_test= ds_cancer_test[['lpsa']]

In [38]:
ds_cancer_trn.head()

lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45      lpsa
0 -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0 -0.430783
1 -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0 -0.162519
2 -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20 -0.162519
3 -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0 -0.162519
4  0.751416  3.432373   62 -1.386294    0 -1.386294        6      0  0.371564

In [39]:
ds_cancer_test.head()

lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45      lpsa
6   0.737164  3.473518   64  0.615186    0 -1.386294        6      0  0.765468
8  -0.776529  3.539509   47 -1.386294    0 -1.386294        6      0  1.047319
9   0.223144  3.244544   63 -1.386294    0 -1.386294        6      0  1.047319
14  1.205971  3.442019   57 -1.386294    0 -0.430783        7      5  1.398717
21  2.059239  3.501043   60  1.474763    0  1.348073        7     20  1.658228

In [40]:
cancer_rgr = linear_model.LinearRegression()
cancer_rgr.fit(x_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [41]:
print("coefficient {}".format(cancer_rgr.coef_))
print("intercept {}".format(cancer_rgr.intercept_))

LR_MSE = mean_squared_error(y_test, cancer_rgr.predict(x_test))
print('Mean squared error for liner regression: {}'.format(LR_MSE))

coefficient [[ 0.57654319  0.61402    -0.01900102  0.14484808  0.73720864 -0.20632423
  -0.02950288  0.00946516]]
intercept [ 0.42917013]
Mean squared error for liner regression: 0.521274005508


In [42]:
import statsmodels.formula.api as sm

model = sm.ols(formula='lcp ~ lcavol + lweight + age + lbph + svi + lcp + gleason + pgg45', data=ds_cancer_test)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                    lcp   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 6.268e+30
Date:                Wed, 07 Oct 2015   Prob (F-statistic):          3.47e-317
Time:                        21:56:15   Log-Likelihood:                 996.78
No. Observations:                  30   AIC:                            -1976.
Df Residuals:                      21   BIC:                            -1963.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept   2.442e-15   4.32e-15      0.566      0.5

In [49]:
ds_cancer_test.corr()

lcavol   lweight       age      lbph       svi       lcp   gleason  \
lcavol   1.000000  0.203455  0.096055 -0.076500  0.398673  0.636998  0.453406   
lweight  0.203455  1.000000  0.526248  0.495874  0.072752  0.205169  0.173297   
age      0.096055  0.526248  1.000000  0.543672  0.079422  0.049455  0.092624   
lbph    -0.076500  0.495874  0.543672  1.000000  0.043540  0.175657  0.170479   
svi      0.398673  0.072752  0.079422  0.043540  1.000000  0.683579  0.356348   
lcp      0.636998  0.205169  0.049455  0.175657  0.683579  1.000000  0.593699   
gleason  0.453406  0.173297  0.092624  0.170479  0.356348  0.593699  1.000000   
pgg45    0.309566  0.257143  0.230829  0.379522  0.394816  0.582940  0.782083   
lpsa     0.737401  0.233263  0.041092 -0.044297  0.599771  0.707463  0.437303   

            pgg45      lpsa  
lcavol   0.309566  0.737401  
lweight  0.257143  0.233263  
age      0.230829  0.041092  
lbph     0.379522 -0.044297  
svi      0.394816  0.599771  
lcp      0.582940  0.707463  
gleason  0.782083  0.437303  
pgg45    1.000000  0.367904  
lpsa     0.367904  1.000000

In [43]:
x_train=ds_cancer_trn[['lcavol','lweight','age','lbph','svi','lcp','gleason','pgg45']]
y_train=ds_cancer_trn[['lpsa']]
x_test = ds_cancer_test[['lcavol','lweight','age','lbph','svi','lcp','gleason','pgg45']]
y_test= ds_cancer_test[['lpsa']]

knn = neighbors.KNeighborsRegressor()
knn.fit(x_train,y_train)
y_ = knn.predict(x_test)

#y_ = knn.fit(x_train.values,y_train).predict(x_test.values)
#plt.scatter(x_train.values,y_train.values,c='k', label='data')
#plt.plot(x_test,y_,c='b',label='prediction')

KNN_MSE = mean_squared_error(y_test, knn.predict(x_test))
print('Mean squared error for KNN: {}'.format(KNN_MSE))

Mean squared error for KNN: 1.20880435689


In [44]:
#X_train_scaled, X_test_scaled, y_train_scaled, y_test_scaled = cross_validation.train_test_split(preprocessing.scale(ds_cancer_trn), ds_cancer_test, test_size=0.2, random_state=0)
x_train=ds_cancer_trn[['lcavol','lweight','age','lbph','svi','lcp','gleason','pgg45']]
y_train=ds_cancer_trn[['lpsa']]
x_test = ds_cancer_test[['lcavol','lweight','age','lbph','svi','lcp','gleason','pgg45']]
y_test= ds_cancer_test[['lpsa']]

X_train_scaled=preprocessing.scale(x_train)
X_test_scaled=preprocessing.scale(x_test)
y_train_scaled=preprocessing.scale(y_train)
y_test_scaled = preprocessing.scale(y_test)

In [45]:
param_grid = [{'alpha':np.linspace(1e-8, 1, 3000)}]

#Ridge
ridge_rgr = linear_model.Ridge(normalize = True)
ridge_cv = grid_search.GridSearchCV(ridge_rgr, param_grid,cv=5)
ridge_cv.fit(X_train_scaled, y_train_scaled)


GridSearchCV(cv=5, error_score='raise',
       estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=True, solver='auto', tol=0.001),
       fit_params={}, iid=True, loss_func=None, n_jobs=1,
       param_grid=[{'alpha': array([  1.00000e-08,   3.33454e-04, ...,   9.99667e-01,   1.00000e+00])}],
       pre_dispatch='2*n_jobs', refit=True, score_func=None, scoring=None,
       verbose=0)

In [46]:
#Lasso
lasso_rgr = linear_model.Lasso(normalize = True)
lasso_cv = grid_search.GridSearchCV(lasso_rgr, param_grid, cv=5)
lasso_cv.fit(X_train_scaled, y_train_scaled)

GridSearchCV(cv=5, error_score='raise',
       estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=True, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False),
       fit_params={}, iid=True, loss_func=None, n_jobs=1,
       param_grid=[{'alpha': array([  1.00000e-08,   3.33454e-04, ...,   9.99667e-01,   1.00000e+00])}],
       pre_dispatch='2*n_jobs', refit=True, score_func=None, scoring=None,
       verbose=0)

In [47]:
Ridge_MSE = mean_squared_error(y_test_scaled, ridge_cv.best_estimator_.predict(X_test_scaled))
Lasso_MSE = mean_squared_error(y_test_scaled, lasso_cv.best_estimator_.predict(X_test_scaled))

print('Mean squared error for Ridge regression: {}'.format(Ridge_MSE))
print('Mean squared error for Lasso regression: {}'.format(Lasso_MSE))

Mean squared error for Ridge regression: 0.509778011955
Mean squared error for Lasso regression: 0.492411948696


In [48]:
from sklearn.metrics import make_scorer, mean_squared_error
mse = make_scorer(mean_squared_error, greater_is_better = False)

param_grid = [{'alpha':np.linspace(1e-8, 1, 3000)}]

#Ridge
ridge_rgr = linear_model.Ridge()
ridge_cv = grid_search.GridSearchCV(ridge_rgr, param_grid,cv=5, scoring=mse)
ridge_cv.fit(X_train_scaled, y_train_scaled)

#Lasso
lasso_rgr = linear_model.Lasso()
lasso_cv = grid_search.GridSearchCV(lasso_rgr, param_grid, cv=5, scoring=mse)
lasso_cv.fit(X_train_scaled, y_train_scaled)


Ridge_MSE = mean_squared_error(y_test_scaled, ridge_cv.best_estimator_.predict(X_test_scaled))
Lasso_MSE = mean_squared_error(y_test_scaled, lasso_cv.best_estimator_.predict(X_test_scaled))

print('Mean squared error for Ridge regression: {}'.format(Ridge_MSE))
print('Mean squared error for Lasso regression: {}'.format(Lasso_MSE))

Mean squared error for Ridge regression: 0.491127417991
Mean squared error for Lasso regression: 0.491127392749


### Questions 

####2. Were there any features that were highly correlated?
    svi	lcp	gleason are highly correlated.
####3. What features were the most important? Unimportant?
    Important : lcoval; weight;gleason
    Unimportant: age
####4. What model worked best? What was the MSE of each?
    Lasso Regression seems to be the model that worked best.
    Mean squared error for liner regression: 0.521274005508
    Mean squared error for KNN: 1.20880435689
    Mean squared error for Ridge regression: 0.509778011955
    Mean squared error for Lasso regression: 0.492411948696
####5. What difficulties did you have with this data set?
   Breaking the data apart into train and test based on a field value versus random selection to detemrine which model works the best.